### PR TITLE
WIP: Fix `--enable-shared` build

### PR DESCRIPTION
**Pull request by Blackhex**
The remaining changes to fix `--enable-shared` build that needs to be put on top of #15.

Tested by https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/7900746928 and https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/7900209529.

### CODE REVIEW

File: gcc/config.gcc [hunk 1]
```diff
@@ -1264,24 +1264,40 @@ aarch64*-*-linux*)
...
+	fi
 	tm_file="${tm_file} aarch64/mingw32.h"
-	tm_file="${tm_file} aarch64/mingw-stdint.h"
-	tmake_file="${tmake_file} aarch64/t-aarch64" # aarch64/t-aarch64-linux"
-# 	tm_defines="${tm_defines}  TARGET_DEFAULT_ASYNC_UNWIND_TABLES=1"
 	user_headers_inc_next_post="${user_headers_inc_next_post} float.h"
+	user_headers_inc_next_pre="${user_headers_inc_next_pre} stddef.h stdarg.h"
+	tm_file="${tm_file} i386/mingw-w64.h"
+	tm_file="${tm_file} aarch64/mingw-stdint.h"
+	tmake_file="${tmake_file} t-winn aarch64/t-cygming t-slibgcc"
```
Comment (Blackhex) [inline gcc/config.gcc]: The main part of the fix is adding `t-slibgcc` which enables `ENABLE_SHARED_LIBGCC` which enables proper linker configuration in `gcc/config/aarch64/mingw32.h`

File: gcc/config.gcc [hunk 2]
```diff
@@ -1264,24 +1264,40 @@ aarch64*-*-linux*)
...
-# 	tm_file="${tm_file} aarch64/aarch64-elf.h aarch64/aarch64-errata.h aarch64/aarch64-linux.h"
-# 	tm_file="${tm_file} i386/unix.h i386/bsd.h i386/gas.h dbxcoff.h i386/cygming.h"	
-	tm_file="${tm_file} aarch64/aarch64-coff.h"
-	tm_file="${tm_file} aarch64/cygming.h"
+	tm_file="${tm_file} aarch64/aarch64-coff.h aarch64/cygming.h" # i386/unix.h i386/bsd.h i386/gas.h
+	xm_file=i386/xm-mingw32.h
+	c_target_objs="${c_target_objs} winnt-c.o"
+	cxx_target_objs="${cxx_target_objs} winnt-c.o"
+	d_target_objs="${d_target_objs} winnt-d.o"
+	target_has_targetcm="yes"
```
Comment (eukarpov) [inline gcc/config.gcc]: Can we move config.cc refactoring to a separate PR?

File: libgcc/config.host [hunk 3]
```diff
@@ -397,10 +397,10 @@ case ${host} in
   *-*-solaris2*)
   enable_execute_stack=enable-execute-stack-mprotect.c
   ;;
-i[34567]86-*-mingw* | x86_64-*-mingw*)
+i[34567]86-*-mingw* | x86_64-*-mingw* | aarch64-*-mingw*)
```
Comment (eukarpov) [inline libgcc/config.host]: I think it might be replaced with `*-mingw*`

File: libgcc/config.host [hunk 4]
```diff
@@ -397,10 +397,10 @@ case ${host} in
...
   enable_execute_stack=config/i386/enable-execute-stack-mingw32.c
   ;;
-i[34567]86-*-cygwin* | x86_64-*-cygwin*)
+i[34567]86-*-cygwin* | x86_64-*-cygwin* | aarch64-*-cygwin*)
```
Comment (eukarpov) [inline libgcc/config.host]: I think it might be replaced with `*-cygwin*`

File: libgcc/config.host [hunk 5]
```diff
@@ -459,29 +459,31 @@ aarch64*-*-vxworks7*)
...
 	  tmake_dlldir_file="i386/t-dlldir-x"
 	fi
-	tmake_file="${tmake_file} ${tmake_eh_file} ${tmake_thr_file} ${tmake_dlldir_file}" # i386/t-chkstk
-        tmake_file="${tmake_file} i386/t-slibgcc-cygming i386/t-slibgcc-mingw i386/t-cygming i386/t-mingw32 t-dfprules"
+	tmake_file="${tmake_file} ${tmake_eh_file} ${tmake_thr_file} ${tmake_dlldir_file} i386/t-slibgcc-cygming i386/t-slibgcc-mingw i386/t-cygming i386/t-mingw32 t-dfprules t-crtfm" # aarch64/t-chkstk
 	tmake_file="${tmake_file} ${cpu_type}/t-aarch64"
-	tmake_file="${tmake_file} ${cpu_type}/t-lse t-slibgcc-libgcc"
-	tmake_file="${tmake_file} ${cpu_type}/t-softfp t-softfp t-crtfm"
+	tmake_file="${tmake_file} ${cpu_type}/t-lse"
+	tmake_file="${tmake_file} ${cpu_type}/t-sme ${cpu_type}/t-softfp t-softfp"
```
Comment (eukarpov) [inline libgcc/config.host]: twice t-softfp?

Comment (Blackhex) [reply libgcc/config.host]: One resolves to `config/aarch64/t-softfp` the other to `config/t-softfp`.

File: gcc/config.gcc (hunk 2 above)
Comment (Blackhex) [reply gcc/config.gcc]: Moved to #15.

File: libgcc/config.host (hunk 3 above)
Comment (Blackhex) [reply libgcc/config.host]: Changed.

File: libgcc/config.host (hunk 4 above)
Comment (Blackhex) [reply libgcc/config.host]: Changed.

File: libgcc/config.host (hunk 5 above)
Comment (Blackhex) [reply libgcc/config.host]: Moved to #15.

**Review by eukarpov — DISMISSED** [review]

**Comment by eukarpov** [comment]
Could it be merged directly to woarm64 and it fixes the problem?

**Comment by Blackhex** [comment]
> Could it be merged directly to woarm64 and it fixes the problem?

No, the most significant parts of the fix are in the https://github.com/Windows-on-ARM-Experiments/gcc-woarm64/pull/15. Unfortunatelly, it does not make sense to separate those fixes from the refactor.

**Comment by eukarpov** [comment]
Then it does not make sense to move to a separate PR.

File: gcc/config.gcc (hunk 1 above)
Comment (eukarpov) [reply gcc/config.gcc]: I would add only changes which are needed for enabled-shared fix

**Comment by Blackhex** [comment]
Closing the PR as it now diverges from the upstream a lot.

**Pull request closed without merging**